### DISCUSSION TITLE
DllImport and ARM

### DISCUSSION THREAD

**Discussion by The-Futurist** [dotnet/runtime#99192] [General] [answered]
I'm working on some embedded stuff, I can run full .Net 8 code on the board and debug it no issues. I want to do some interop so I installed the tools for ARM:

![image](https://github.com/dotnet/runtime/assets/12262952/4ed953df-a0a6-40c3-be5b-747d314264c6)

My problem now is that no matter what I do, Visual Studio seems to insist on generating a COFF output file (with .DLL suffix) and there's no way I can get the compiler/linker to generate it in ELF format (usually named .SO files).

At runtime, no matter what I do, I get a runtime exception:

`/root/.epdata/EndpointTesting/NativeLibrary.dll: invalid ELF header`

It finds the file but is expecting it to be an ELF format rather than COFF, is this hard coded into the runtime? Is there no way to be able to specify the format as COFF/ELF?

Is this part of .Net or is it deeper down, perhaps part of Linux?

Is there no way to have Visual Studio's ARM compiler create an ELF formatted .SO file?

**Comment by huoyaoyuan** [comment, marked as answer]
Architecture (x86, x86-64 or ARM) is orthogonol with OS (Windows or Linux). You are creating binaries for ARM64 version of Windows. Since your device runs Linux, you need Linux ARM toolchain.
